### ISSUE TITLE
measured boot eventlog parsing fails with tpm2-tools v5.6

### ISSUE THREAD

**Issue by kkaarreell** [keylime/keylime#1496]
Is your issue a feature request? If so, please raise it as an [enhancement](https://github.com/keylime/enhancements)

# Environment
 * OS / version:  Fedora-Rawhide
 * Processor architecture:  x86_64
 * TPM Manufacturer:  swtpm
 * Keylime version: current upstream

# Description 
In recent CI runs there are failures in  /functional/measured-boot-swtpm-sanity  on Fedora Rawhide.
https://artifacts.dev.testing-farm.io/7e150df8-b4d8-4aa9-b1cc-7bb756d51a2d/
According to my testing this has been caused by the update from tpm2-tools-5.5-5.fc40 to tpm2-tools-5.6-1.fc40 which have introduced some changes in the tpm2_evenlog parser.
In the verifier log there is the following error:
```
ov 15 12:39:16 ip-172-31-23-57.us-east-2.compute.internal keylime_verifier[76390]: 2023-11-15 12:39:16.726 - keylime.measured_boot - ERROR - Boot attestation failed for agent d432fbb3-d2f1-4a97-9ef7-75bd81c00000, policy example, refstate={"has_secureboot": true, "scrtm_and_bios": [{"scrtm": {"sha256": "0x96a296d224f285c67bee93c30f8a309157f0daa35dc5b87e410b78630a09cfc7"}, "platform_firmware": [{"sha256": "0xdf32b6fc4e8a153713feba580b9ca6da6efa8cb23dd907f0e543d1f4c5be3c14"}, {"sha256": "0xda36c15e1184901ad53ec231d441224d2fff5ff168d8972fe71efca4e150b502"}]}], "pk": [{"SignatureOwner": "8be4df61-93ca-11d2-aa0d-00e098032b8c", "SignatureData": "0x308203a030820288a003020102020900fef588e8f396c0f1300d06092a864886f70d01010b05003051312b302906035504031322526564204861742053656375726520426f6f742028504b2f4b454b206b65792031293122302006092a864886f70d0109011613736563616c657274407265646861742e636f6d301e170d3134313033313131313533375a170d3337313032353131313533375a3051312b302906035504031322526564204861742053656375726520426f6f742028504b2f4b454b206b65792031293122302006092a864886f70d0109011613736563616c657274407265646861742e636f6d30820122300d06092a864886f70d01010105000382010f003082010a0282010100901f847b8dbceb9726826d88ab8ac98c6870f9df4b07b237830b02c86768309ee3f0f0994ab85957c641f6388bfe664c49e93737922e98011e5b1450e6a88d250df586e6ab30cb4016ea8d8b1686704337f2cec091df71148e990e89b64c6d241e8ce42f4f25d0ba06f8c6e8191876731d816da8d805cf3ac87b28c836a3160d298c999a68dcabc04d8dbf5abb2ba9394b04971cf936bbc53a8604aeafd4827be0abde490568fcf6ae681a6c904d57193c646603f6c7529bf794cf936aa168c9aacf996bbcaa5e08e7391cf7f80fba067ef1cbe876ddfe22daad3a5e5b34eab3c9e04d04297eb860b905efb5d91758561660b93032f0364ac3f2798d124070f30203010001a37b307930090603551d1304023000302c06096086480186f842010d041f161d4f70656e53534c2047656e657261746564204365727469666963617465301d0603551d0e041604143ce960e3ff19a10a7ba342f48d422eb4d59c72ec301f0603551d230418301680143ce960e3ff19a10a7ba342f48d422eb4d59c72ec300d06092a864886f70d01010b050003820101005c4d9288b4825f1dad8b11ecdf06a67aa52b9f37550c8d6e0500adb70c418969cfd665069b5178d2adc7bf9cdc05737fe71e3913b4eab6307d4075ab9c430bdfb0c21bbf30e0f4fec0db622198f6c5afde3b4f490ae61ef986b03f0dd6d44637db54745eff11c260c67058c51c6fecb2d86e6fc3bc338738a4f344649c343b28942678279f1617e83b690a25a973367e9e375cece83fdb91f912b33dcee7dd15c3ae8c0520619b95de9baffab15c1ce597e7c3341185f58a2726a47036ec0cf6833d90f736f3f9f315d49062be53b4afd349afeff473e87b76e4442a37ba81a4990c3a312471a0e4e4b71acb47e4aa22cfef756180e343b7485773113d789b69"}], "kek": [{"SignatureOwner": "a0baa8a3-041d-48a8-bc87-c36d121b5e3d", "SignatureData": "0x308203a030820288a003020102020900fef588e8f396c0f1300d06092a864886f70d01010b05003051312b302906035504031322526564204861742053656375726520426f6f742028504b2f4b454b206b65792031293122302006092a864886f70d0109011613736563616c657274407265646861742e636f6d301e170d3134313033313131313533375a170d3337313032353131313533375a3051312b302906035504031322526564204861742053656375726520426f6f742028504b2f4b454b206b65792031293122302006092a864886f70d0109011613736563616c657274407265646861742e636f6d30820122300d06092a864886f70d01010105000382010f003082010a0282010100901f847b8dbceb9726826d88ab8ac98c6870f9df4b07b237830b02c86768309ee3f0f0994ab85957c641f6388bfe664c49e93737922e98011e5b1450e6a88d250df586e6ab30cb4016ea8d8b1686704337f2cec091df71148e990e89b64c6d241e8ce42f4f25d0ba06f8c6e8191876731d816da8d805cf3ac87b28c836a3160d298c999a68dcabc04d8dbf5abb2ba9394b04971cf936bbc53a8604aeafd4827be0abde490568fcf6ae681a6c904d57193c646603f6c7529bf794cf936aa168c9aacf996bbcaa5e08e7391cf7f80fba067ef1cbe876ddfe22daad3a5e5b34eab3c9e04d04297eb860b905efb5d91758561660b93032f0364ac3f2798d124070f30203010001a37b307930090603551d1304023000302c06096086480186f842010d041f161d4f70656e53534c2047656e657261746564204365727469666963617465301d0603551d0e041604143ce960e3ff19a10a7ba342f48d422eb4d59c72ec301f0603551d230418301680143ce960e3ff19a10a7ba342f48d422eb4d59c72ec300d06092a864886f70d01010b050003820101005c4d9288b4825f1dad8b11ecdf06a67aa52b9f37550c8d6e0500adb70c418969cfd665069b5178d2adc7bf9cdc05737fe71e3913b4eab6307d4075ab9c430bdfb0c21bbf30e0f4fec0db622198f6c5afde3b4f490ae61ef986b03f0dd6d44637db54745eff11c260c67058c51c6fecb2d86e6fc3bc338738a4f344649c343b28942678279f1617e83b690a25a973367e9e375cece83fdb91f912b33dcee7dd15c3ae8c0520619b95de9baffab15c1ce597e7c3341185f58a2726a47036ec0cf6833d90f736f3f9f315d49062be53b4afd349afeff473e87b76e4442a37ba81a4990c3a312471a0e4e4b71acb47e4aa22cfef756180e343b7485773113d789b69"}, {"SignatureOwner": "77fa9abd-0359-4d32-bd60-28f4e78f784b", "SignatureData": "0x308205e8308203d0a003020102020a610ad188000000000003300d06092a864886f70d01010b0500308191310b3009060355040613025553311330110603550408130a57617368696e67746f6e3110300e060355040713075265646d6f6e64311e301c060355040a13154d6963726f736f667420436f72706f726174696f6e313b3039060355040313324d6963726f736f667420436f72706f726174696f6e205468697264205061727479204d61726b6574706c61636520526f6f74301e170d3131303632343230343132395a170d3236303632343230353132395a308180310b3009060355040613025553311330110603550408130a57617368696e67746f6e3110300e060355040713075265646d6f6e64311e301c060355040a13154d6963726f736f667420436f72706f726174696f6e312a3028060355040313214d6963726f736f667420436f72706f726174696f6e204b454b204341203230313130820122300d06092a864886f70d01010105000382010f003082010a0282010100c4e8b58abfad5726b026c3eae7fb577a44025d070dda4ae5742ae6b00fec6debec7fb9e35a63327c11174f0ee30ba73815938ec6f5e084b19a9b2ce7f5b791d609e1e2c004a8ac301cdf48f306509a64a7517fc8854f8f2086cefe2fe19fff82c0ede9cdcef4536a623a0b43b9e225fdfe05f9d4c414ab11e223898d70b7a41d4decaee59cfa16c2d7c1cbd4e8c42fe599ee248b03ec8df28beac34afb4311120b7eb547926cdce60489ebf53304eb10012a71e5f983133cff25092f687646ffba4fbedcad712a58aafb0ed2793de49b653bcc292a9ffc7259a2ebae92eff6351380c602ece45fcc9d76cdef6392c1af79408479877fe352a8e89d7b07698f150203010001a382014f3082014b301006092b06010401823715010403020100301d0603551d0e0416041462fc43cda03ea4cb6712d25bd955ac7bccb68a5f301906092b0601040182371402040c1e0a00530075006200430041300b0603551d0f040403020186300f0603551d130101ff040530030101ff301f0603551d2304183016801445665243e17e5811bfd64e9e2355083b3a226aa8305c0603551d1f045530533051a04fa04d864b687474703a2f2f63726c2e6d6963726f736f66742e636f6d2f706b692f63726c2f70726f64756374732f4d6963436f725468695061724d6172526f6f5f323031302d31302d30352e63726c306006082b0601050507010104543052305006082b060105050730028644687474703a2f2f7777772e6d6963726f736f66742e636f6d2f706b692f63657274732f4d6963436f725468695061724d6172526f6f5f323031302d31302d30352e637274300d06092a864886f70d01010b05000382020100d48488f514941802ca2a3cfb2a921c0cd7a0d1f1e85266a8eea2b5757a9000aa2da4765aea79b7b9376a517b1064f6e164f20267bef7a81b78bdbace8858640cd657c819a35f05d6dbc6d069ce484b32b7eb5dd230f5c0f5b8ba7807a32bfe9bdb345684ec82caae4125709c6be9fe900fd7961fe5e7941fb22a0c8d4bff2829107bf7d77ca5d176b905c879ed0f90929cc2fedf6f7e6c0f7bd4c145dd345196390fe55e56d8180596f407a642b3a077fd0819f27156cc9f8623a487cba6fd587ed4696715917e81f27f13e50d8b8a3c8784ebe3cebd43e5ad2d84938e6a2b5a7c44fa52aa81c82d1cbbe052df0011f89a3dc160b0e133b5a388d165190a1ae7ac7ca4c182874e38b12f0dc514876ffd8d2ebc39b6e7e6c3e0e4cd2784ef9442ef298b9046413b811b67d8f9435965cb0dbcfd00924ff4753ba7a924fc50414079e02d4f0a6a27766e52ed96697baf0ff78705d045c2ad5314811ffb3004aa373661da4a691b34d868edd602cf6c940cd3cf6c2279adb1f0bc03a24660a9c407c22182f1fdf2e8793260bfd8aca522144bcac1d84beb7d3f5735b2e64f75b4b060032253ae91791dd69b411f15865470b2de0d350f7cb03472ba97603bf079eba2b21c5da216b887c5e91bf6b597256f389fe391fa8a7998c3690eb7a31c200597f8ca14ae00d7c4f3c01410756b34a01bb59960f35cb0c5574e36d23284bf9e"}], "db": [{"SignatureOwner": "77fa9abd-0359-4d32-bd60-28f4e78f784b", "SignatureData": "0x308205d7308203bfa003020102020a61077656000000000008300d06092a864886f70d01010b0500308188310b3009060355040613025553311330110603550408130a57617368696e67746f6e3110300e060355040713075265646d6f6e64311e301c060355040a13154d6963726f736f667420436f72706f726174696f6e31323030060355040313294d6963726f736f667420526f6f7420436572746966696361746520417574686f726974792032303130301e170d3131313031393138343134325a170d3236313031393138353134325a308184310b3009060355040613025553311330110603550408130a57617368696e67746f6e3110300e060355040713075265646d6f6e64311e301c060355040a13154d6963726f736f667420436f72706f726174696f6e312e302c060355040313254d6963726f736f66742057696e646f77732050726f64756374696f6e20504341203230313130820122300d06092a864886f70d01010105000382010f003082010a0282010100dd0cbba2e42e09e3e7c5f79669bc0021bd693333efad04cb5480ee0683bbc52084d9f7d28bf338b0aba4ad2d7c627905ffe34a3f04352070e3c4e76be09cc03675e98a31dd8d70e5dc37b5744696285b8760232cbfdc47a567f751279e72eb07a6c9b91e3b53357ce5d3ec27b9871cfeb9c923096fa84691c16e963c41d3cba33f5d026a4dec691f25285c36fffd43150a94e019b4cfdfc212e2c25b27ee2778308b5b2a096b22895360162cc0681d53baec49f39d618c85680973445d7da2542bdd79f715cf355d6c1c2b5ccebc9c238b6f6eb526d93613c34fd627aeb9323b41922ce1c7cd77e8aa544ef75c0b048765b44318a8b2e06d1977ec5a24fa48030203010001a38201433082013f301006092b06010401823715010403020100301d0603551d0e04160414a92902398e16c49778cd90f99e4f9ae17c55af53301906092b0601040182371402040c1e0a00530075006200430041300b0603551d0f040403020186300f0603551d130101ff040530030101ff301f0603551d23041830168014d5f656cb8fe8a25c6268d13d94905bd7ce9a18c430560603551d1f044f304d304ba049a0478645687474703a2f2f63726c2e6d6963726f736f66742e636f6d2f706b692f63726c2f70726f64756374732f4d6963526f6f4365724175745f323031302d30362d32332e63726c305a06082b06010505070101044e304c304a06082b06010505073002863e687474703a2f2f7777772e6d6963726f736f66742e636f6d2f706b692f63657274732f4d6963526f6f4365724175745f323031302d30362d32332e637274300d06092a864886f70d01010b0500038202010014fc7c7151a579c26eb2ef393ebc3c520f6e2b3f101373fea868d048a6344d8a960526ee3146906179d6ff382e456bf4c0e528b8da1d8f8adb09d71ac74c0a36666a8cec1bd70490a81817a49bb9e240323676c4c15ac6bfe404c0ea16d3acc368ef62acdd546c503058a6eb7cfe94a74e8ef4ec7c867357c2522173345af3a38a56c804da0709edf88be3cef47e8eaef0f60b8a08fb3fc91d727f53b8ebbe63e0e33d3165b081e5f2accd16a49f3da8b19bc242d090845f541dff89eaba1d47906fb0734e419f409f5fe5a12ab21191738a2128f0cede73395f3eab5c60ecdf0310a8d309e9f4f69685b67f51886647198da2b0123d812a680577bb914c627bb6c107c7ba7a8734030e4b627a99e9cafcce4a37c92da4577c1cfe3ddcb80f5afad6c4b30285023aeab3d96ee4692137de81d1f675190567d393575e291b39c8ee2de1cde445735bd0d2ce7aab1619824658d05e9d81b367af6c35f2bce53f24e235a20a7506f6185699d4782cd1051bebd088019daa10f105dfba7e2c63b7069b2321c4f9786ce2581706362b911203cca4d9f22dbaf9949d40ed1845f1ce8a5c6b3eab03d370182a0a6ae05f47d1d5630a32f2afd7361f2a705ae5425908714b57ba7e8381f0213cf41cc1c5b990930e88459386e9b12099be98cbc595a45d62d6a0630820bd7510777d3df345b99f979fcb57806f33a904cf77a4621c597e"}, {"SignatureOwner": "77fa9abd-0359-4d32-bd60-28f4e78f784b", "SignatureData": "0x30820610308203f8a003020102020a6108d3c4000000000004300d06092a864886f70d01010b0500308191310b3009060355040613025553311330110603550408130a57617368696e67746f6e3110300e060355040713075265646d6f6e64311e301c060355040a13154d6963726f736f667420436f72706f726174696f6e313b3039060355040313324d6963726f736f667420436f72706f726174696f6e205468697264205061727479204d61726b6574706c61636520526f6f74301e170d3131303632373231323234355a170d3236303632373231333234355a308181310b3009060355040613025553311330110603550408130a57617368696e67746f6e3110300e060355040713075265646d6f6e64311e301c060355040a13154d6963726f736f667420436f72706f726174696f6e312b3029060355040313224d6963726f736f667420436f72706f726174696f6e2055454649204341203230313130820122300d06092a864886f70d01010105000382010f003082010a0282010100a5086c4cc745096a4b0ca4c0877f06750c43015464e0167f07ed927d0bb273bf0c0ac64a4561a0c5162d96d3f52ba0fb4d499b4180903cb954fde6bcd19dc4a4188a7f418a5c59836832bb8c47c9ee71bc214f9a8a7cff443f8d8f32b22648ae75b5eec94c1e4a197ee4829a1d78774d0cb0bdf60fd316d3bcfa2ba551385df5fbbadb7802dbffec0a1b96d583b81913e9b6c07b407be11f2827c9faef565e1ce67e947ec0f044b27939e5dab2628b4dbf3870e2682414c933a40837d558695ed37cedc1045308e74eb02a876308616f631559eab22b79d70c61678a5bfd5ead877fba86674f71581222042222ce8bef547100ce503558769508ee6ab1a201d50203010001a382017630820172301206092b060104018237150104050203010001302306092b060104018237150204160414f8c16bb77f77534af325371d4ea1267b0f207080301d0603551d0e0416041413adbf4309bd82709c8cd54f316ed522988a1bd4301906092b0601040182371402040c1e0a00530075006200430041300b0603551d0f040403020186300f0603551d130101ff040530030101ff301f0603551d2304183016801445665243e17e5811bfd64e9e2355083b3a226aa8305c0603551d1f045530533051a04fa04d864b687474703a2f2f63726c2e6d6963726f736f66742e636f6d2f706b692f63726c2f70726f64756374732f4d6963436f725468695061724d6172526f6f5f323031302d31302d30352e63726c306006082b0601050507010104543052305006082b060105050730028644687474703a2f2f7777772e6d6963726f736f66742e636f6d2f706b692f63657274732f4d6963436f725468695061724d6172526f6f5f323031302d31302d30352e637274300d06092a864886f70d01010b05000382020100350842ff30cccef7760cad1068583529463276277cef124127421b4aaa6d813848591355f3e95834a6160b82aa5dad82da808341068fb41df203b9f31a5d1bf15090f9b3558442281c20bdb2ae5114c5c0ac9795211c90db0ffc779e95739188cabdbd52b905500ddf579ea061ed0de56d25d9400f1740c8cea34ac24daf9a121d08548fbdc7bcb92b3d492b1f32fc6a21694f9bc87e4234fc3606178b8f2040c0b39a257527cdc903a3f65dd1e736547ab950b5d312d107bfbb74dfdc1e8f80d5ed18f42f14166b2fde668cb023e5c784d8edeac13382ad564b182df1689507cdcff072f0aebbdd8685982c214c332bf00f4af06887b592553275a16a826a3ca32511a4edadd704aecbd84059a084d1954c6291221a741d8c3d470e44a6e4b09b3435b1fab653a82c81eca40571c89db8bae81b4466e447540e8e567fb39f1698b286d0683e9023b52f5e8f50858dc68d825f41a1f42e0de099d26c75e4b669b52186fa07d1f6e24dd1daad2c77531e253237c76c52729586b0f135616a19f5b23b815056a6322dfea289f94286271855a182ca5a9bf830985414a64796252fc826e441941a5c023fe596e3855b3c3e3fbb47167255e22522b1d97be703062aa3f71e9046c3000dd61989e30e352762037115a6efd027a0a0593760f83894b8e07870f8ba4c868794f6e0ae0245ee65c2b6a37e69167507929bf5a6bc598358"}], "dbx": [{"SignatureOwner": "a0baa8a3-041d-48a8-bc87-c36d121b5e3d", "SignatureData": "0xe3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"}], "mokdig": [], "mokxdig": [], "kernels": [{"shim_authcode_sha256": "0x8c5806e66bb5b052ebf860e1722474269cff3dde588610df21dbe8cf12c08390", "grub_authcode_sha256": "0x3ca0250a57e52ec2d8eedc0674fdfaba376bf1f0297483dfe63bc3abb40b073d", "kernel_authcode_sha256": "0xb7aa67ab83a8ebe76393e0cf8ba25f9e7b5dc734740cf87e68640f391c207732", "initrd_plain_sha256": "0x9103a58e6e52a7ec7a598a0f387d246e195deaa2014824d43e4bf2757796dd00", "kernel_cmdline": "\\(hd0,gpt2\\)/vmlinuz\\-5\\.14\\.10\\-300\\.fc35\\.x86_64\\ root=UUID=cb32c90d\\-89aa\\-43ed\\-9a28\\-14a5452cd091\\ ro\\ console=ttyS0,115200\\ ima_appraise=fix\\ ima_policy=tcb\\ ima_template=ima\\-sig\u0000"}]}, reason={'EventNum': 26, 'PCRIndex': 14, 'EventType': 'EV_IPL', 'DigestCount': 1, 'Digests': [{'AlgorithmId': 'sha256', 'Digest': 'bdc8aa461f5b498d4619090d647888ae9c442e966883e78d8b52f4e3881165e1'}], 'EventSize': 8, 'Event': {'String': 'MokList\x00'}} [Event String is not 'MokList', Event String is not 'MokListX', Event String is not 'MokListTrusted']
```
See `MokList\x00` while previously we had `MokList\0`. I guess keylime should cut-off the trailing '\x00' like it previously did for '\0'.

## Expected behavior vs. actual behavior 
measured boot attestation test passes

## Steps to reproduce problem
1. schedule the above mentioned test

**Comment by JuergenReppSIT** [comment]
@kkaarreell There was a tpm2 tool PR which did fix yaml string formatting:
https://github.com/tpm2-software/tpm2-tools/pull/3270
Now there is another issue related to this topic:
https://github.com/tpm2-software/tpm2-tools/issues/3303
Would the suggested surrounding the hex values with double quotes also cause problems for you?

**Comment by kkaarreell** [comment]
> @kkaarreell There was a tpm2 tool PR which did fix yaml string formatting: [tpm2-software/tpm2-tools#3270](https://github.com/tpm2-software/tpm2-tools/pull/3270) Now there is another issue related to this topic: [tpm2-software/tpm2-tools#3303](https://github.com/tpm2-software/tpm2-tools/issues/3303) Would the suggested surrounding the hex values with double quotes also cause problems for you?

Thank you for these details.  I think I would rather ask @THS-on to answer that.

**Comment by williamcroberts** [comment]
Didn't Keylime move to rust? What does it still use the tools for? Just for event log parsing?

**Comment by kkaarreell** [comment]
Just the agent is in Rust. Python component Is using tpm2-tools to read the binary eventlog.

**Comment by kkaarreell** [comment]
Yes

**Comment by maugustosilva** [comment]
Right, the very last dependency on `tpm2-tools` inside the `verifier` is one invocation of `tpm2_eventlog --eventlog-version=2 <log_bin_filename>`. We actually have a "drop-in replacement" ready to go (https://github.com/keylime/python3-uefi-eventlog), written by @galmasi. It should happen first quarter of 2024.
